# Laboratorio 2 - Parte 2

### Modelos no paramétricos

### 2018-II

#### Profesor: Julián D. Arias Londoño
#### julian.ariasl@udea.edu.co


## Guía del laboratorio

En esta archivo va a encontrar tanto celdas de código cómo celdas de texto con las instrucciones para desarrollar el laboratorio.

Lea atentamente las instrucciones entregadas en las celdas de texto correspondientes y proceda con la solución de las preguntas planteadas.

Nota: no olvide ir ejecutando las celdas de código de arriba hacia abajo para que no tenga errores de importación de librerías o por falta de definición de variables.

In [2]:
#Tomado de https://github.com/rramosp/mooc-grader
from Autentication import *#python 3
import inspect, urllib
html, auth_code, userinfo = google_authenticate(PORT_NUMBER=8080)
html

Haga click en el siguiente enlace para autenticarse con su cuenta de correo institucional
https://accounts.google.com/o/oauth2/auth?scope=profile+email&client_id=893762525034-g9d91ddls9e19a1q77c7hsq2rhgqo9h7.apps.googleusercontent.com&access_type=offline&redirect_uri=http%3A%2F%2Flocalhost%3A8080&response_type=code
waiting for authentication ...
authentication succeeded
/?code=4/TgDA9tXMcGU1db2RtY5ZmPyRPl8zOG6eG6AZXx_dTVzVE4ZHSrs_8jskpfyp9C2Is4jTneVOIgMt0tWbRccmm2c


#### Segundo integrante:

David Alejandro Marín Alzate

In [3]:
%matplotlib inline
import numpy as np
import scipy as sc
import matplotlib.pyplot as plt
from __future__ import division
import pixiedust

#Algunas advertencias que queremos evitar
import warnings
warnings.filterwarnings("always")

Pixiedust database opened successfully


## Ejercicio 1

Para el problema de clasificación usaremos la base de datos iris, que cargaremos desde la librería sklearn. Responda las siguientes preguntas:

In [4]:
#cargamos la bd iris desde el dataset de sklearn
from sklearn import datasets

iris = datasets.load_iris()

#Esta es la base de datos Iris del UCI Machine Learning Repository 
#https://archive.ics.uci.edu/ml/datasets/iris

X, Y = iris.data, iris.target


/home/sofia/simulacion/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/sofia/simulacion/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/sofia/simulacion/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/sofia/simulacion/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)


In [5]:
print("Muestras de entrada ", X.shape)
print("Diferente clases que hay en la salida ",np.unique(Y))
print("Muestras x clase ", np.unique(Y, return_counts=True))

Muestras de entrada  (150, 4)
Diferente clases que hay en la salida  [0 1 2]
Muestras x clase  (array([0, 1, 2]), array([50, 50, 50]))


1.1 ¿Cu&aacute;ntas clases tiene el problema?: 

La base de datos cuenta con **3 clases**, en este caso correponde a [0,1,2]

1.2 ¿Cu&aacute;ntas caracter&iacute;sticas tiene el problema?:

La base de datos de Iris cuenta con **4 carácteristicas**.

1.3 ¿Cu&aacute;ntas muestras tiene el problema?:

El problema presenta con **150 muestras**.

1.4 ¿Cu&aacute;ntas muestras por cada clase hay en la base de datos?:

Por cada clase hay **50 muestras**, se observa que la distribuición es equitativa.

<b>Nota</b>: Agregue una celda con el código usado para responder cada una de las preguntas anteriores.

## Ejercicio 2

Analice los siguientes métodos a la luz de la teoría vista para los modelos de K vecinos más cercanos (KNN) y complete el código del método KNN.

<b>Nota</b>: Para el cáculo de la distancia entre vectores tienen dos opciones, usar la función scipy.spatial.distance.euclidean o usar la función numpy.linalg.norm. <b>Revisen la documentación </b> para comprender su uso. También serán de utilidad las funciones numpy.sort y numpy.argsort.

In [39]:
from scipy.stats import mode

import operator

#Para calcular el error en los problemas de clasificación
def ErrorClas(Y_lest, Y):
    error = 1 - np.sum(Y_lest == Y)/len(Y)
    
    return error


def KNN(X_train, Y_train, X_val, k, tipo):
    
    #X_train es la matriz con las muestras de entrenamiento
    #Y_train es un vector con los valores de salida pra cada una de las muestras de entrenamiento
    #X_val es la matriz con las muestras de validación
    #tipo es una bandera que indica si el problema es de regresión o de clasificación.
    tipo = tipo
    k = k    #Parámetro k que equivale al número de vecinos a tener en cuenta para resolver el problema de
    #predicción de la variable de salida
    
    N = np.shape(Y_train)[0]
    Y_val = np.zeros(N).reshape(N,1)
    Y_train = Y_train.reshape(N,1) #Asigna espacio correctamente de Y
    
    # obtenemos la matriz con todas las distancias de las 1500 muestras nuevas
    # con respecto a todas las 3500 muestras de entrenamiento
    D = distance(X_val, X_train)
    D_argsort = np.argsort(D, axis = 1)
    
    #print("***********D_argsort")
    #print(D_argsort[0:10, : ])   
    
    Nt = len(X_val)
 
    #Matriz con la salida i correspondiente a la posición indicada en D_argsort
    Y_argsort = Yout(k, D_argsort, Y_train)

    Y_val = np.zeros(Nt).T
    

    if(tipo == 0):
        #regresion
        for x in range (Nt):
            average = np.average(Y_argsort[x])
            Y_val[x] = average
            
    elif(tipo == 1):
        #clasificacion
        
        for x in range (Nt):
            mode = stats.mode(Y_argsort[x])           
            Y_val[x] = mode[0].astype(float)
            
    Y_val.reshape(Nt,1)
    # print(Y_val, Y_val.shape)
    #¿por qué no muestra forma correcta del vector?
    
    return Y_val #Debe retornar un vector que contenga las predicciones para cada una de las muestras en X_val, en el mismo orden.  

def distance(X_val, X_train):
    
    L = np.dot(X_val, X_train.T)
    tem = np.sum(X_val * X_val, axis=1).reshape([X_val.shape[0],1])
    tem2 = np.sum(X_train * X_train, axis=1).reshape([1,X_train.shape[0]])
    
    D = np.dot(tem, np.ones((1,X_train.shape[0]))) + np.dot(np.ones((X_val.shape[0],1)), tem2) - 2* L
    D = np.sqrt(D)
    
    return D

#Me retorna la salida Y_i proveniente de la matriz que contiene la posición de la distancia más cortas (vecinos más cercanos)
# de las nuevas muestras
def Yout (k, D_argsort, Y_train):
    N = len(D_argsort)    
    kn = D_argsort[ : , 0: k]
    kn = kn.reshape(len(D_argsort),k)
    Y_val = np.zeros((N,k))

  #  Y_val =np.array( [[Y_train[kn[y,x]] for x in range(k)] for y in range(N)])
    for x in range (N):
        for y in range (k):
               Y_val[x][y] = Y_train[kn[x,y]]
    
    return Y_val

## Ejercicio 3

Una vez haya completado el codigo del método de KNN, ejecute varias veces el proceso de entrenamiento y evaluacion cambiando el parametro $k$, el cual es el numero de vecinos, y complete la siguiente tabla con los valores del error de clasificación obtenidos:

In [174]:
from numpy import random
from sklearn.model_selection import train_test_split
import math

def mainKNN(k_vecinos, tipoKNN = 1):
    N = np.size(X,0)
    #Validamos el modelo
    random.seed(1)
    Error = np.zeros(4)
    for j in range(4):

        Xtrain, Xtest, Ytrain, Ytest = train_test_split(X, Y, test_size=0.25)

        #Normalizamos los datos
        media = np.mean(Xtrain)
        desvia = np.std(Xtrain)
        Xtrain = sc.stats.stats.zscore(Xtrain)
        Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

        Yest = KNN(Xtrain, Ytrain, Xtest, k_vecinos, tipoKNN)#Complete con el llamado apropiado de la función que Uds diseñaron

        #Evaluamos las predicciones del modelo con los datos de test
        Error[j] = ErrorClas(Yest,Ytest)
    print("\n************************* Número vecinos: ", k_vecinos)
    print('Error durante la prueba = ' + str(np.mean(Error)) + '+-' + str(np.std(Error)))

main(2)


************************* Número vecinos:  2
Error durante la prueba = 0.26973684210526316+-0.08802031684381349


In [175]:
print("Muestras entrenamiento: ",Xtrain.shape)
print("Muestras validación",len(Xtest))

Muestras entrenamiento:  (114, 4)
Muestras validación 36


Responda:

3.1 ¿Qué metodología de validación se usa en el experimento?

La metodología de validación implementada es **validación cruzada**, es este caso $k = 4$  folds.

3.2 ¿Cuántas muestras se usan para el entrenamiento?

Hay **112 muestras** para entrenar

3.3 ¿Cuántas muestras se usan para la validación?

En general son **36 muestras** nuevas.

In [176]:
k_vecinos = [mainKNN(x) for x in range(1,8) ]


************************* Número vecinos:  1
Error durante la prueba = 0.20394736842105263+-0.11376063464335907

************************* Número vecinos:  2
Error durante la prueba = 0.26973684210526316+-0.08802031684381349

************************* Número vecinos:  3
Error durante la prueba = 0.26973684210526316+-0.08603090020146066

************************* Número vecinos:  4
Error durante la prueba = 0.2763157894736842+-0.0842516347030638

************************* Número vecinos:  5
Error durante la prueba = 0.26315789473684215+-0.09304036594559835

************************* Número vecinos:  6
Error durante la prueba = 0.2763157894736842+-0.0842516347030638

************************* Número vecinos:  7
Error durante la prueba = 0.26973684210526316+-0.08802031684381349


#### Tabla de resultados

In [72]:
import pandas as pd
import qgrid
randn = np.random.randn
df_types = pd.DataFrame({
    'Numero de vecinos' : pd.Series(['1', '2', '3', '4', '5', '6', '7'])})
df_types["Error_Prueba"] = ""
df_types["Desviación estándar del error"] = ""
df_types.set_index(['Numero de vecinos'], inplace=True)
df_types["Error_Prueba"][1] = "0.2697"
df_types["Desviación estándar del error"][1] = "0.088"
#df_types.sort_index(inplace=True)
qgrid_widget = qgrid.show_grid(df_types, show_toolbar=False)
qgrid_widget

/home/sofia/simulacion/lib/python3.5/site-packages/ipywidgets/widgets/widget.py:281: DeprecationWarning: Widget registration using a string name has been deprecated. Widget registration now uses a plain `@register` decorator.
  warnings.warn("Widget registration using a string name has been deprecated. Widget registration now uses a plain `@register` decorator.", DeprecationWarning)
/home/sofia/simulacion/lib/python3.5/site-packages/qgrid/grid.py:573: DeprecationWarning: metadata {'sync': True} was set from the constructor. With traitlets 4.1, metadata should be set using the .tag() method, e.g., Int().tag(key1='value1', key2='value2')
  _df_json = Unicode('', sync=True)
/home/sofia/simulacion/lib/python3.5/site-packages/qgrid/grid.py:833: DeprecationWarning: QgridWidget._precision_changed is deprecated in traitlets 4.1: use @observe and @unobserve instead.
  def _precision_changed(self):


QgridWidget(grid_options={'highlightSelectedCell': False, 'enableColumnReorder': False, 'autoEdit': False, 'ed…

Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [73]:
qgrid_widget.get_changed_df()

Error_Prueba Desviación estándar del error
Numero de vecinos                                           
1                       0.2039                        0.1137
2                       0.2697                         0.088
3                       0.2697                         0.086
4                       0.2763                        0.0842
5                       0.2631                         0.093
6                       0.2763                        0.0842
7                       0.2697

## Ejercicio 4

A continuación debe completar el código del método parzen para resolver problemas de clasificación con el modelo ventana de Parzen.

In [169]:
#Calcula el kernel gaussiano de x
def kernel_gaussiano(x):
    return np.exp((-0.5)*x**2)

def ParzenWindow(x,Data,h):
    h = h
    Ns = Data.shape[0]
    suma = 0
    for k in range(Ns):
        u = sc.spatial.distance.euclidean(x,Data[k,:])
        suma += kernel_gaussiano(u/h)
    return suma

def ParzenClass(X_train, Y_train, X_val, ancho_h):
    #Tamaño muestras nuevas
    Nx = X_val.shape[0]
    Yest = np.zeros(Nx)
    h = ancho_h
    
    #Separ las clases
    clase1 = X_train[Y_train==0]
    clase2 = X_train[Y_train == 1]
    clase3 = X_train[Y_train == 2]
    
    p1= np.zeros(Nx)
    p2= np.zeros(Nx)   
    p3= np.zeros(Nx)
    
    for i in range(Nx):        
        p1[i] = ParzenWindow(X_val[i],clase1, h)
        p2[i] = ParzenWindow(X_val[i],clase2, h)
        p3[i] = ParzenWindow(X_val[i],clase3, h)        
        p = np.array([p1[i], p2[i], p3[i]])
        clase_max = np.argmax(p)
        Yest[i] = clase_max 
    
    return Yest#Debe retornar un vector que contenga las predicciones para cada una de las muestras en X_val, en el mismo orden.  

## Ejercicio 5

Ahora debe resolver el mismo problema de clasificación con el método ventana de Parzen. Complete el código ejecute las pruebas para diferentes valores de $h$ y llene la siguiente tabla:

In [190]:
from numpy import random
from sklearn.model_selection import StratifiedKFold
import math

def mainPW(h=0.9):
    #Validamos el modelo
    random.seed(1)
    Error = np.zeros(4)
    skf = StratifiedKFold(n_splits=4)
    j = 0
    for train, test in skf.split(X, Y):
        Xtrain = X[train,:]
        Ytrain = Y[train]
        Xtest = X[test,:]
        Ytest = Y[test]
        #Normalizamos los datos
        media = np.mean(Xtrain)
        desvia = np.std(Xtrain)
        Xtrain = sc.stats.stats.zscore(Xtrain)
        Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

        Yest = ParzenClass(Xtrain, Ytrain, Xtest,h)#Complete con el llamado apropiado de la función que Uds diseñaron
     
        #Evaluamos las predicciones del modelo con los datos de test
        Error[j] = ErrorClas(Yest,Ytest)
        j += 1
    print("\n**************************** Ancho ventana ", h)
    print('Error durante la prueba = ' + str(np.mean(Error)) + '+-' + str(np.std(Error)))

In [191]:
print("Muestras entrenamiento ",Xtrain.shape)
print("Muestras para validación ",Xtest.shape)

Muestras entrenamiento  (114, 4)
Muestras para validación  (36, 4)


Responda:
    
5.1 ¿Qué metodología de validación se usó en la simulación?:

**Validación cruzada estratificada** con $k = 4$ splits, valida que las particiones aleatorias contenga proporcialmente valores de cada categoría


5.2 ¿Cuántas muestras se usan para el entrenamiento?:

Se destinan *114 muestras* para entrenamiento


5.3 ¿Cuántas muestras se usan para la validación?:

Se cuenta con *36 muestras* para validación

In [192]:
ancho_h = [0.05, 0.1, 0.5, 1, 2, 5, 10 ]
pw = [mainPW(x) for x in ancho_h]


**************************** Ancho ventana  0.05
Error durante la prueba = 0.2216880341880342+-0.04332854434165311

**************************** Ancho ventana  0.1
Error durante la prueba = 0.2216880341880342+-0.04332854434165311

**************************** Ancho ventana  0.5
Error durante la prueba = 0.281517094017094+-0.04021339704891218

**************************** Ancho ventana  1
Error durante la prueba = 0.280982905982906+-0.03145818774430656

**************************** Ancho ventana  2
Error durante la prueba = 0.24198717948717952+-0.058466142137675314

**************************** Ancho ventana  5
Error durante la prueba = 0.2216880341880342+-0.04696907336459245

**************************** Ancho ventana  10
Error durante la prueba = 0.2216880341880342+-0.04696907336459245


#### Tabla de resultados

In [181]:
import pandas as pd
import qgrid
randn = np.random.randn
df_types2 = pd.DataFrame({
    'Tamano de ventana' : pd.Series(['0.05', '0.1', '0.5', '1', '2', '5', '10'])})
df_types2["Error_Prueba"] = ""
df_types2["Desviación estándar del error"] = ""
df_types2.set_index(['Tamano de ventana'], inplace=True)
df_types2["Error_Prueba"][1] = "0.222"
df_types2["Desviación estándar del error"][1] = "0.043"
#df_types.sort_index(inplace=True)
qgrid_widget = qgrid.show_grid(df_types2, show_toolbar=False)
qgrid_widget

QgridWidget(grid_options={'highlightSelectedCell': False, 'enableColumnReorder': False, 'autoEdit': False, 'ed…

Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [182]:
qgrid_widget.get_changed_df()

Error_Prueba Desviación estándar del error
Tamano de ventana                                           
0.05                    0.2216                         0.043
0.1                      0.222                         0.043
0.5                     0.2815                        0.0402
1                        0.242                        0.0314
2                        0.242                        0.0584
5                       0.2216                        0.0469
10                      0.2216                        0.0469

¿Por qué cuando el ancho de ventana es 5 y 10 da el mismo error de prueba y desviación estándar?
